src/jenkins.ts

```diff
@@ -0,0 +1,228 @@
+    const metadata = {
+      namespace: props.metadata?.namespace ?? 'default',
+      labels: props.metadata?.labels ?? { app: 'jenkins' },
+    };
+    const authorizationStrategy = 'createUser';
+    const disableCSRFProtection = props.disableCsrfProtection ?? false;
+    // Keeping container resources defaults similar to Jenkins defaults here: https://jenkinsci.github.io/kubernetes-operator/docs/getting-started/latest/installing-the-operator/
+    const containers = [
+      {
+        name: 'jenkins-master',
+        image: 'jenkins/jenkins:lts',
+        imagePullPolicy: 'Always',
+        resources: {
+          limits: {
+            cpu: jenkins.JenkinsSpecMasterContainersResourcesLimits.fromString('1500m'),
+            memory: jenkins.JenkinsSpecMasterContainersResourcesLimits.fromString('3Gi'),
+          },
+          requests: {
+            cpu: jenkins.JenkinsSpecMasterContainersResourcesLimits.fromString('1'),
+            memory: jenkins.JenkinsSpecMasterContainersResourcesLimits.fromString('500M'),
+          },
+        },
+      },
+    ];;
+
+    const jenkinsCustomResource = new jenkins.Jenkins(this, 'Default', {
+      metadata: metadata,
+      spec: {
+        groovyScripts: {
+          configurations: [],
+          secret: {
+            name: '',
+          },
+        },
+        configurationAsCode: {
+          configurations: [],
+          secret: {
+            name: '',
+          },
+        },
+        jenkinsApiSettings: {
+          authorizationStrategy: authorizationStrategy,
+        },
+        master: {
+          disableCsrfProtection: disableCSRFProtection,
+          containers: containers,
+          basePlugins: this._basePlugins,
+          plugins: this._plugins,
+        },
+        seedJobs: this._seedJobs,
+      },
+    });
+
+    this.node.defaultChild = jenkinsCustomResource;
+  }
+
+  /**
+   * Add base plugins to jenkins instance.
+   * @param basePlugins List of base plugins.
+   */
+  public addBasePlugins(...basePlugins: Plugin[]): void {
+    const inputBasePlugins = [...basePlugins];
+    const existingPlugins = [...this._basePlugins];
+
+    for (const basePlugin of inputBasePlugins) {
+      for (const existingPlugin of existingPlugins) {
+        if (existingPlugin.name == basePlugin.name && existingPlugin.version != basePlugin.version) {
+          const indexOfExistingPlugin = this._basePlugins.findIndex((plugin) => {
+            return plugin.name == existingPlugin.name;
+          });
+          this._basePlugins.splice(indexOfExistingPlugin, 1);
+          const updatedPlugin = {
+            name: basePlugin.name,
+            version: basePlugin.version,
+          };
+          this._basePlugins.push(updatedPlugin);
+
+          const indexOfInputPlugin = basePlugins.findIndex((plugin) => {
+            return plugin.name == basePlugin.name;
+          });
+          basePlugins.splice(indexOfInputPlugin, 1);
+        }
+      }
+    }
```

Comment:
What if instead, you store plugins internally as a map? where the keys are the names. Then at synth time, convert the map to an array. 

Feels like it's going to greatly simplify the implementation. 

Use `Lazy` for delayed synthesis. See https://github.com/cdk8s-team/cdk8s-plus/blob/k8s-24/main/src/service.ts#L218

Reply:
Makes sense, taking a look. Thank you for recommending :)